connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecutePlanResponseSender.scala

```diff
@@ -0,0 +1,111 @@
+
+/**
+ * ExecutePlanResponseSender sends responses to the GRPC stream.
+ * It runs on the RPC thread, and gets notified by ExecutePlanResponseObserver about available
+ * responses.
+ * It notifies the ExecutePlanResponseObserver back about cached responses that can be removed
+ * after being sent out.
+ * @param responseObserver the GRPC request StreamObserver
+ */
+private[connect] class ExecutePlanResponseSender(
+  executionObserver: ExecutePlanResponseObserver,
+  grpcObserver: StreamObserver[ExecutePlanResponse]) {
+
+  executionObserver.setExecutePlanResponseSender(this)
+
+  private val signal = new Object
+  private var detached = false
+
+  def notifyResponse(): Unit = {
+    signal.synchronized {
+      signal.notify()
+    }
+  }
+
+  def detach(): Unit = {
+    signal.synchronized {
+      this.detached = true
+      signal.notify()
+    }
+  }
+
+  /**
+   * Receive responses from executionObserver and send them to grpcObserver.
+   * @param lastSentIndex Start sending the stream from response after this.
+   * @return true if finished because stream completed
+   *         false if finished because stream was detached
+   */
+  def run(lastSentIndex: Long): Boolean = {
+    // register to be notified about available responses.
+    executionObserver.setExecutePlanResponseSender(this)
+
+    var currentIndex = lastSentIndex + 1
+    var finished = false
+
+    while (!finished) {
+      var response: Option[CachedExecutePlanResponse] = None
+      // Get next available response.
+      // Wait until either this sender got detached or next response is ready,
+      // or the stream is complete and it had already sent all responses.
+      signal.synchronized {
+        while (!detached && response.isEmpty &&
+            executionObserver.getLastIndex().forall(currentIndex <= _)) {
+          response = executionObserver.getResponse(currentIndex)
+          // If response is empty, wait to get notified.
+          // We are cholding signal here, so:
+          // - if detach() is waiting on signal, and will acquire it after wait()
+          //   here releases it, and wait() will be waked up by notify.
+          // - if getLastIndex() or getResponse() changed, executionObserver would call
+          //   notify(), which would wait on signal, and will acquire it after wait() here releases
+          //   it, and wait will be waked up by notify.
+          if (response.isEmpty) {
+            signal.wait()
+          }
+        }
+      }
+
+      // Send next available response.
+      if (detached) {
+        // This sender got detached by the observer.
+        finished = true
+      } else if (response.isDefined) {
+        // There is a response available to be sent.
+        grpcObserver.onNext(response.get.r)
+        // Remove after sending.
+        executionObserver.removeUntilIndex(currentIndex)
+        currentIndex += 1
+      } else if (executionObserver.getLastIndex().forall(currentIndex > _)) {
```

Comment:
```suggestion
      } else if (executionObserver.getLastIndex().forall(currentIndex >= _)) {
```

?

Reply:
At least if I'm not wrong if you have a one batch response first `currentIndex=0` which is incremented after fetching it, then its `currentIndex=1` but now, the lastindex will be 1 as well and currentIndex is not > then lastIndex